### PR TITLE
Tools: support schema for custom POJO parameters

### PR DESCRIPTION
**Pull request by gs-snagaraj**
Support POJOs as `@Tool` method parameters.
Repurpose `@Description` to be used to describe POJO fields.

Example:

```java
@Data
static class Query {

    @Description("List of fields to fetch records")
    List<String> select;

    @Description("List of conditions to filter on. Pass null if no condition")
    List<Condition> where;

    @Description("limit on number of records")
    Integer limit;

    @Description("offset for fetching records")
    Integer offset;
}

@Data
static class Condition {

    @Description("Field to filter on")
    String field;

    @Description("Operator to apply")
    Operator operator;

    @Description("Value to compare with")
    Object value;
}

enum Operator {

    EQUALS,
    NOT_EQUALS,
    IS_NULL,
    IS_NOT_NULL
}

@Tool
public String executeQuery(Query query) {
    ....
}
```

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 1]
```diff
@@ -98,7 +95,72 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
             return removeNulls(STRING, enums((Object[]) type.getEnumConstants()), description);
         }
 
-        return removeNulls(OBJECT, description); // TODO provide internals
+        return removeNulls(OBJECT, schema(type), description);
+    }
+
+    private static JsonSchemaProperty schema(Class<?> structured){
+        //TODO: Externalize the depth
+        return schema(structured, new HashSet<>(), 3);
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: Why depth is 3?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 2]
```diff
@@ -98,7 +95,72 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
...
+    private static JsonSchemaProperty schema(Class<?> structured, Set<Class<?>> visited, int level) {
+        if (level < 0 || visited.contains(structured)) {
+            return null;
+        }
+
+        visited.add(structured);
+        Map<String,Object> properties = new HashMap<>();
+        for (Field field : structured.getDeclaredFields()) {
+            String name = field.getName();
+            if (name.equals("__$hits$__") || name.equals("this$0") || java.lang.reflect.Modifier.isStatic(field.getModifiers())) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: what is `this$0`?
could you pelase add a comment?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 3]
```diff
@@ -98,7 +95,72 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
...
+            Map<Object,Object> objectMap = new HashMap<>();
+            for(JsonSchemaProperty jsonSchemaProperty : schemaProperties) {
+                objectMap.put(jsonSchemaProperty.key(), jsonSchemaProperty.value());
+            }
+            properties.put(name, objectMap);
+        }
+        return from( "schema", properties );
+    }
+
+    static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Field field, Set<Class<?>> visited , int level) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: there is a lot of duplicated logic copy-pasted from `toJsonSchemaProperties(Parameter parameter)`, can this be extracted and reused?

**Review by ghost — COMMENTED** [review]
@gs-snagaraj thanks a lot!

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 1 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: This is to keep a check on how deep we recursively generate schema. 
Wanted this be configurable, but not sure how to do in this framework

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 3 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: Done

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 2 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: done, inner classes would be represent as a field `this$0` we need to skip that.

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: Ah right, thanks  for explanation!

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 1 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: In theory, the object graph can be deeper than 3 levels. Recursion is stopped by using the map of visited classes. Do we really need this depth checks?

**Review by gs-snagaraj — COMMENTED** [review]
Done. Removed the depth logic

**Comment by gs-snagaraj** [comment]
@langchain4j Could you please review and merge this.

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 4]
```diff
@@ -75,6 +72,69 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
...
+        Iterable<JsonSchemaProperty> simpleType = toJsonSchemaProperties(type, description);
+
+        if (simpleType != null) {
+            return simpleType;
+        }
+
+        if (Collection.class.isAssignableFrom(type)) {
+            return removeNulls(ARRAY, arrayTypeFrom(parameter.getParameterizedType()), description);
+        }
+
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:95]: just do double check: do we need to have a similar logic for collections here (similar to array)?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 5]
```diff
@@ -75,6 +72,69 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
...
+        Iterable<JsonSchemaProperty> simpleType = toJsonSchemaProperties(type, description);
+
+        if (simpleType != null) {
+            return simpleType;
+        }
+
+        if (Collection.class.isAssignableFrom(type)) {
+            return removeNulls(ARRAY, arrayTypeFrom((Class<?>) ((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0]), description);
+        }
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:142]: just do double check: do we need to have a similar logic for arrays here (similar to collection)?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 6]
```diff
@@ -94,17 +154,15 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
         if (type.isArray()) {
             return removeNulls(ARRAY, arrayTypeFrom(type.getComponentType()), description);
         }
-        if (Collection.class.isAssignableFrom(type)) {
-            return removeNulls(ARRAY, arrayTypeFrom(parameter.getParameterizedType()), description);
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:109]: just do double check: was this removal on purpose?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 7]
```diff
@@ -94,17 +154,15 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
...
-        }
 
         if (type.isEnum()) {
             return removeNulls(STRING, enums((Class<?>) type), description);
         }
 
-        return removeNulls(OBJECT, description); // TODO provide internals
+        return null;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:173]: in which case can this happen?

**Review by ghost — COMMENTED** [review]
@gs-snagaraj thank you so much for your contribution!

This seems to be a quite complicated logic you are changing in this PR, so I kindly ask you to do thorough testing and hope for your understading.

Could you please add more unit (`ToolSpecificationsTest`) and integration (`AiServicesWithToolsIT`) tests for such cases:
- recursion: root POJO referencing another POJO of the same type (e.g. `Person` has `Person parent`)
- tree of POJOs: root POJO referencing another POJO which references yet another POJO
- root POJO argument contains a collection (list/set) of primitives
- root POJO contains a collection (list/set) of POJOs
- root POJO contains an array of primitives
- root POJO contains an array of POJOs
- @Tool has a not a single root POJO as an argument, but a collection (list/set) of such POJOs
- @Tool has an array of such POJOs as an argument

Thank you so much in advance!

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 4 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:95]: Yes. Array is considered SimpleType and handled in ` toJsonSchemaProperties(Class<?> type, JsonSchemaProperty description)`

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 5 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:142]: Yes. Array is considered SimpleType and handled in  toJsonSchemaProperties(Class<?> type, JsonSchemaProperty description)

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 6 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:109]: Yes. Collection is considered Complex Type and handled in  `toJsonSchemaProperties(Parameter parameter)` and `toJsonSchemaProperties(Field field, Set<Class<?>> visited)`

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 7 above)
Comment (gs-snagaraj) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:173]: All SimpleType are handled in `toJsonSchemaProperties(Class<?> type, JsonSchemaProperty description)`

If its Complex type we written null, so that it can be handled in ` toJsonSchemaProperties(Parameter parameter)` and `toJsonSchemaProperties(Field field, Set<Class<?>> visited)`

**Comment by gs-snagaraj** [comment]
@langchain4j Have covered the requested changes in `ToolSpecificationsTest`

Please check and merge.

**Comment by geoand** [comment]
Can you please explain the rationale for this? Can you show a few examples of what this brings compares to the current state?

Thanks

**Comment by G0dC0der** [comment]
> Can you please explain the rationale for this? Can you show a few examples of what this brings compares to the current state?
> 
> Thanks

Currently, only basic types are allowed(string, int, arrays etc) are allowed as method parameter for tools. If I am not mistaken, this feature expand the allowed types with POJOs. 

As someone who uses this framework, I strongly welcome this feature!

**Comment by gs-snagaraj** [comment]
With This, we could make ai build an Object and pass as parameter.

Ex: Assume we have queryService which takes following Query Object
```
public class Query{
@Description("select fields for query")
List<String> select;
@Description("list of condition for where clause")
List<Condition> where;
@Description("limit on no of records to fetch")
Integer limit;
@Description("no of records to skip")
Integer skip
}

public class Condition{
@Description("field")
Sring field;
@Description("operator")
String operator
@Description("value")
Object value;
}
```

Now we want an AI tool which can generate this object and call the tool

```
      @Tool("returns rows from  database for a given query")
       List<Result> getTransactionAmount(@P("query") Query  query) {

            return queryService.getResults(query);
           
        }
```

With the above changes, we Tool specification will generate a schema of the object, and AI can generate the json, which can be serialised to Query object

**Comment by langchain4j** [comment]
HI @gs-snagaraj does it actually work for you?

I have tested the following code and it does not work:

```java
public class Test {

    static OpenAiChatModel model = OpenAiChatModel.builder()
            .apiKey(System.getenv("OPENAI_API_KEY"))
            .temperature(0.0)
            .logRequests(true)
            .logResponses(true)
            .build();

    interface Assistant {

        String chat(String message);
    }

    static class Person {

        String name;
        String surname;
        int age;
    }

    static class Tools {

        @Tool
        void validate(Person person) {
            System.out.println(person.name + " " + person.surname + " " + person.age);
        }
    }

    public static void main(String[] args) {

        Assistant assistant = AiServices.builder(Assistant.class)
                .chatLanguageModel(model)
                .tools(new Tools())
                .build();

        assistant.chat("Validate John Doe");
    }
}
```

Here are the logs:

```
2024-06-13 10:32:18 [main] dev.ai4j.openai4j.RequestLoggingInterceptor.logDebug()
DEBUG: Request:
- method: POST
- url: http://langchain4j.dev:8082/v1/chat/completions
- headers: [Authorization: Bearer sk-FS...Y5], [User-Agent: langchain4j-openai]
- body: {
  "model": "gpt-3.5-turbo",
  "messages": [
    {
      "role": "user",
      "content": "Validate John Doe"
    }
  ],
  "temperature": 0.0,
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "validate",
        "description": "",
        "parameters": {
          "type": "object",
          "properties": {
            "arg0": {
              "schema": {
                "surname": {
                  "type": "string"
                },
                "name": {
                  "type": "string"
                },
                "age": {
                  "type": "integer"
                }
              },
              "type": "object"
            }
          },
          "required": [
            "arg0"
          ]
        }
      }
    }
  ]
}
2024-06-13 10:32:19 [main] dev.ai4j.openai4j.ResponseLoggingInterceptor.logDebug()
DEBUG: Response:
- status code: 200
- headers: [Server: nginx/1.18.0], [Date: Thu, 13 Jun 2024 08:32:20 GMT], [Content-Type: application/json], [Transfer-Encoding: chunked], [Connection: keep-alive], [openai-organization: liubarskyi], [openai-processing-ms: 272], [openai-version: 2020-10-01], [strict-transport-security: max-age=15724800; includeSubDomains], [x-ratelimit-limit-requests: 5000], [x-ratelimit-limit-tokens: 160000], [x-ratelimit-remaining-requests: 4999], [x-ratelimit-remaining-tokens: 159978], [x-ratelimit-reset-requests: 12ms], [x-ratelimit-reset-tokens: 8ms], [x-request-id: req_f266cdb4a249c38c62ddf077a393c812], [CF-Cache-Status: DYNAMIC], [CF-RAY: 8930b3d0becc4527-ATL], [alt-svc: h3=":443"; ma=86400], [X-Cache-Status: HIT]
- body: {
  "id": "chatcmpl-9Za5wvcKevav9npr3SUjQhOrlKmlP",
  "object": "chat.completion",
  "created": 1718267256,
  "model": "gpt-3.5-turbo-0125",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": null,
        "tool_calls": [
          {
            "id": "call_Vv4fp4uz01GAkyJnXn4HIeij",
            "type": "function",
            "function": {
              "name": "validate",
              "arguments": "{}"
            }
          }
        ]
      },
      "logprobs": null,
      "finish_reason": "tool_calls"
    }
  ],
  "usage": {
    "prompt_tokens": 33,
    "completion_tokens": 9,
    "total_tokens": 42
  },
  "system_fingerprint": null
}

2024-06-13 10:32:19 [main] dev.langchain4j.agent.tool.DefaultToolExecutor.execute()
DEBUG: About to execute ToolExecutionRequest { id = "call_Vv4fp4uz01GAkyJnXn4HIeij", name = "validate", arguments = "{}" } for memoryId default
2024-06-13 10:32:19 [main] dev.ai4j.openai4j.RequestLoggingInterceptor.logDebug()
DEBUG: Request:
- method: POST
- url: http://langchain4j.dev:8082/v1/chat/completions
- headers: [Authorization: Bearer sk-FS...Y5], [User-Agent: langchain4j-openai]
- body: {
  "model": "gpt-3.5-turbo",
  "messages": [
    {
      "role": "user",
      "content": "Validate John Doe"
    },
    {
      "role": "assistant",
      "content": null,
      "tool_calls": [
        {
          "id": "call_Vv4fp4uz01GAkyJnXn4HIeij",
          "type": "function",
          "function": {
            "name": "validate",
            "arguments": "{}"
          }
        }
      ]
    },
    {
      "role": "tool",
      "tool_call_id": "call_Vv4fp4uz01GAkyJnXn4HIeij",
      "content": "Cannot read field \"name\" because \"person\" is null"
    }
  ],
  "temperature": 0.0,
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "validate",
        "description": "",
        "parameters": {
          "type": "object",
          "properties": {
            "arg0": {
              "schema": {
                "surname": {
                  "type": "string"
                },
                "name": {
                  "type": "string"
                },
                "age": {
                  "type": "integer"
                }
              },
              "type": "object"
            }
          },
          "required": [
            "arg0"
          ]
        }
      }
    }
  ]
}
2024-06-13 10:32:19 [main] dev.langchain4j.agent.tool.DefaultToolExecutor.execute()
ERROR: Error while executing tool: java.lang.NullPointerException: Cannot read field "name" because "person" is null
	at dev.langchain4j.service.Test$Tools.validate(Test.java:35)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at dev.langchain4j.agent.tool.DefaultToolExecutor.execute(DefaultToolExecutor.java:62)
	at dev.langchain4j.agent.tool.DefaultToolExecutor.execute(DefaultToolExecutor.java:43)
	at dev.langchain4j.service.DefaultAiServices$1.invoke(DefaultAiServices.java:179)
	at dev.langchain4j.service.$Proxy3.chat(Unknown Source)
	at dev.langchain4j.service.Test.main(Test.java:46)
2024-06-13 10:32:19 [main] dev.ai4j.openai4j.ResponseLoggingInterceptor.logDebug()
DEBUG: Response:
- status code: 200
- headers: [Server: nginx/1.18.0], [Date: Thu, 13 Jun 2024 08:32:20 GMT], [Content-Type: application/json], [Transfer-Encoding: chunked], [Connection: keep-alive], [openai-organization: liubarskyi], [openai-processing-ms: 398], [openai-version: 2020-10-01], [strict-transport-security: max-age=15724800; includeSubDomains], [x-ratelimit-limit-requests: 5000], [x-ratelimit-limit-tokens: 160000], [x-ratelimit-remaining-requests: 4999], [x-ratelimit-remaining-tokens: 159964], [x-ratelimit-reset-requests: 12ms], [x-ratelimit-reset-tokens: 13ms], [x-request-id: req_1e857c4dda0fab3cad2ad68a029d0819], [CF-Cache-Status: DYNAMIC], [CF-RAY: 8930b3d4abfa1385-ATL], [alt-svc: h3=":443"; ma=86400], [X-Cache-Status: HIT]
- body: {
  "id": "chatcmpl-9Za5xbM28aWGdteib5Y9GHCxIrHWA",
  "object": "chat.completion",
  "created": 1718267257,
  "model": "gpt-3.5-turbo-0125",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "John Doe is not a valid input for validation. Please provide a valid input for validation."
      },
      "logprobs": null,
      "finish_reason": "stop"
    }
  ],
  "usage": {
    "prompt_tokens": 61,
    "completion_tokens": 19,
    "total_tokens": 80
  },
  "system_fingerprint": null
}

```

I am not sure what exactly is wrong, but it seems that instead of `schema` there should be `properties` in the tool definition ?

**Comment by gs-snagaraj** [comment]
@langchain4j  Missed a small change. Now pushed the fix along with a Unit test. 
If the ai passes null, we have error back to ai. Then it will construct Object and send it.

**Comment by gs-snagaraj** [comment]
Sample Logs```
2024-06-14 21:26:56 [main] dev.ai4j.openai4j.RequestLoggingInterceptor.logDebug()
DEBUG: Request:
- method: POST
- url: https://api.openai.com/v1/chat/completions

- body: {
  "model": "gpt-3.5-turbo",
  "messages": [
    {
      "role": "user",
      "content": "List 5 users where country is India"
    }
  ],
  "temperature": 0.0,
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "executeQuery",
        "description": "Execute the query and return the result",
        "parameters": {
          "type": "object",
          "properties": {
            "arg0": {
              "description": "query to execute",
              "type": "object",
              "properties": {
                "select": {
                  "description": "List of fields to fetch records",
                  "type": "array",
                  "items": {
                    "type": "string"
                  }
                },
                "offset": {
                  "description": "offset for fetching records",
                  "type": "integer"
                },
                "limit": {
                  "description": "limit on number of records",
                  "type": "integer"
                },
                "where": {
                  "description": "List of conditions to filter on. Pass null if no condition",
                  "type": "array",
                  "items": {
                    "type": "object",
                    "properties": {
                      "field": {
                        "description": "Field to filter on",
                        "type": "string"
                      },
                      "value": {
                        "description": "Value to compare with",
                        "type": "object",
                        "properties": {}
                      },
                      "operator": {
                        "description": "Operator to apply",
                        "type": "string",
                        "enum": [
                          "EQUALS",
                          "NOT_EQUALS",
                          "IS_NULL",
                          "IS_NOT_NULL"
                        ]
                      }
                    }
                  }
                }
              }
            }
          },
          "required": [
            "arg0"
          ]
        }
      }
    }
  ]
}
2024-06-14 21:26:58 [main] dev.ai4j.openai4j.ResponseLoggingInterceptor.logDebug()
DEBUG: Response:
- status code: 200
- body: {
  "id": "chatcmpl-9a3aLQ5uMupHNgtwy0ahpyjDlHErj",
  "object": "chat.completion",
  "created": 1718380617,
  "model": "gpt-3.5-turbo-0125",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": null,
        "tool_calls": [
          {
            "id": "call_qUTPi4m53EbqS8w8vaJ5K7M2",
            "type": "function",
            "function": {
              "name": "executeQuery",
              "arguments": "{\"arg0\":{\"select\":[\"name\",\"country\"],\"where\":[{\"field\":\"country\",\"value\":\"India\",\"operator\":\"EQUALS\"}],\"limit\":5}}"
            }
          }
        ]
      },
      "logprobs": null,
      "finish_reason": "tool_calls"
    }
  ],
  "usage": {
    "prompt_tokens": 162,
    "completion_tokens": 38,
    "total_tokens": 200
  },
  "system_fingerprint": null
}

2024-06-14 21:26:58 [main] dev.langchain4j.agent.tool.DefaultToolExecutor.execute()
DEBUG: About to execute ToolExecutionRequest { id = "call_qUTPi4m53EbqS8w8vaJ5K7M2", name = "executeQuery", arguments = "{"arg0":{"select":["name","country"],"where":[{"field":"country","value":"India","operator":"EQUALS"}],"limit":5}}" } for memoryId default
query to execute2024-06-14 21:26:58 [main] dev.langchain4j.agent.tool.DefaultToolExecutor.execute()
DEBUG: Tool execution result: [ {name = Amar}, {name= Akbar} ,{ name = Antony} ]
2024-06-14 21:26:58 [main] dev.ai4j.openai4j.RequestLoggingInterceptor.logDebug()
DEBUG: Request:
- method: POST
- url: https://api.openai.com/v1/chat/completions
- body: {
  "model": "gpt-3.5-turbo",
  "messages": [
    {
      "role": "user",
      "content": "List 5 users where country is India"
    },
    {
      "role": "assistant",
      "content": null,
      "tool_calls": [
        {
          "id": "call_qUTPi4m53EbqS8w8vaJ5K7M2",
          "type": "function",
          "function": {
            "name": "executeQuery",
            "arguments": "{\"arg0\":{\"select\":[\"name\",\"country\"],\"where\":[{\"field\":\"country\",\"value\":\"India\",\"operator\":\"EQUALS\"}],\"limit\":5}}"
          }
        }
      ]
    },
    {
      "role": "tool",
      "tool_call_id": "call_qUTPi4m53EbqS8w8vaJ5K7M2",
      "content": "[ {name \u003d Amar}, {name\u003d Akbar} ,{ name \u003d Antony} ]"
    }
  ],
  "temperature": 0.0,
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "executeQuery",
        "description": "Execute the query and return the result",
        "parameters": {
          "type": "object",
          "properties": {
            "arg0": {
              "description": "query to execute",
              "type": "object",
              "properties": {
                "select": {
                  "description": "List of fields to fetch records",
                  "type": "array",
                  "items": {
                    "type": "string"
                  }
                },
                "offset": {
                  "description": "offset for fetching records",
                  "type": "integer"
                },
                "limit": {
                  "description": "limit on number of records",
                  "type": "integer"
                },
                "where": {
                  "description": "List of conditions to filter on. Pass null if no condition",
                  "type": "array",
                  "items": {
                    "type": "object",
                    "properties": {
                      "field": {
                        "description": "Field to filter on",
                        "type": "string"
                      },
                      "value": {
                        "description": "Value to compare with",
                        "type": "object",
                        "properties": {}
                      },
                      "operator": {
                        "description": "Operator to apply",
                        "type": "string",
                        "enum": [
                          "EQUALS",
                          "NOT_EQUALS",
                          "IS_NULL",
                          "IS_NOT_NULL"
                        ]
                      }
                    }
                  }
                }
              }
            }
          },
          "required": [
            "arg0"
          ]
        }
      }
    }
  ]
}
2024-06-14 21:26:59 [main] dev.ai4j.openai4j.ResponseLoggingInterceptor.logDebug()
DEBUG: Response:
- status code: 200
- body: {
  "id": "chatcmpl-9a3aMzmHkVt0pWdPNyGovKjuRHWJL",
  "object": "chat.completion",
  "created": 1718380618,
  "model": "gpt-3.5-turbo-0125",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "Here are 3 users from India:\n1. Amar\n2. Akbar\n3. Antony"
      },
      "logprobs": null,
      "finish_reason": "stop"
    }
  ],
  "usage": {
    "prompt_tokens": 227,
    "completion_tokens": 21,
    "total_tokens": 248
  },
  "system_fingerprint": null
}
```

**Comment by gs-snagaraj** [comment]
log: 
```
{
      "role": "assistant",
      "content": null,
      "tool_calls": [
        {
          "id": "call_qUTPi4m53EbqS8w8vaJ5K7M2",
          "type": "function",
          "function": {
            "name": "executeQuery",
            "arguments": "{\"arg0\":{\"select\":[\"name\",\"country\"],\"where\":[{\"field\":\"country\",\"value\":\"India\",\"operator\":\"EQUALS\"}],\"limit\":5}}"
          }
        }
      ]
    }

File: langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java [hunk 8]
```diff
@@ -48,13 +54,6 @@ static Stream<ChatLanguageModel> models() {
                         .temperature(0.0)
                         .logRequests(true)
                         .logResponses(true)
-                        .build(),
-                MistralAiChatModel.builder()
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:52]: Is there a reason for removing this model?

Comment (gs-snagaraj) [reply langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:52]: I dint had Mistral Locally, so removed. we can add it back.

Comment (langchain4j) [reply langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:52]: That is cloud mistral.

But yes, please bring it back

Comment (gs-snagaraj) [reply langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:52]: Its already there. Some how it doesn't show here. Please merge and check in the main branch

**Review by ghost — APPROVED** [review]
@gs-snagaraj thanks a lot!

**Comment by codexvn** [comment]
whether the required parameter is supported inside the custom POJO parameters？i see 
```
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "validate",
        "description": "",
        "parameters": {
          "type": "object",
          "properties": {
            "arg0": {
              "schema": {
                "surname": {
                  "type": "string"
                },
                "name": {
                  "type": "string"
                },
                "age": {
                  "type": "integer"
                }
              },
              "type": "object"
            }
          },
          "required": [
            "arg0"
          ]
        }
      }
    }
  ]
```
I still want `required` in `properties`

**Comment by langchain4j** [comment]
We could probably add all POJO fields as `required` by default and if the field is `Optional<T>`, then it is not required

**Comment by codexvn** [comment]
> We could probably add all POJO fields as `required` by default and if the field is `Optional<T>`, then it is not required

does this mean that the final result is a standard json schema

**Comment by langchain4j** [comment]
@codexvn I am not sure I understand what you mean. It is already a standart json schema, the proposed change is to include `required` not only in the root of the `parameters`, but also for each nested property
